tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation.

```

Comment:
can we just do this entire validation using a single regex or PEG?

Like I know we don't want to pull dependencies needlessly, but if we can't run the validation script for lack of (good) regex library or python I'm not sure that's a problem.

Reply:
As far as I know, this doesn't easily map to what regexes do. (I haven't used Boost.Spirit so I don't know how well it maps to grammars.)

First, I want to detect multiple problems and complain about them independently, so the user can be told what needs to be fixed. With the current code, I can scan for tabs, disallowed characters, trailing whitespace, and line endings in a single pass, yet diagnose them independently.

Second, at least one of the things that I diagnose isn't really amenable to a regex. I can detect mixed line endings (which occur in practice surprisingly often).

The main deficiency of my current code is that I don't emit context for disallowed characters. Now that I reuse a 64KB buffer, emitting context is not very easy (because it could cross block boundaries).

I'd like to resolve this as "you have a good point but I don't see how to apply it here".

Reply:
yeah, I was more thinking of PEGs or "actually not so regular" regular expressions. something like (<tabs>)?|(...)?.

but yeah, this is not that bad either.